### ISSUE TITLE
[4.1] Full-City Inference Pipeline

### ISSUE THREAD

**Issue by poltergeistjoya** [poltergeistjoya/Vacant_Lot_Detection#21]
## Summary
Build sliding-window inference pipeline to predict vacancy probability across entire NYC. Input: raw NAIP. Output: continuous probability map [0,1], same CRS/resolution as NAIP.

## Key Difference from Training
- **Training:** Uses masking (label=255) to ignore buildings/water
- **Inference:** NO masking—model takes raw NAIP and predicts on ALL pixels

## Implementation Details
- **New file:** `vacant_lot/inference.py`

```python
def predict_full_raster(model_path, naip_vrt_path, output_path, 
                       patch_size=256, overlap=64, batch_size=16):
    '''Sliding-window inference with Gaussian blending.
    
    Overlapping windows + Gaussian weights avoid boundary artifacts.
    For each output pixel: average predictions from overlapping patches.
    '''

def threshold_probability_map(prob_path, output_path, threshold=0.5):
    '''Binary: prob >= threshold → 1, else 0'''

def remove_small_regions(binary_path, output_path, min_area_pixels=50):
    '''Remove isolated predictions < 50 pixels'''
    # Use morphological opening (erosion + dilation)
```

## Inference Parameters
- **Overlap:** 64 pixels (1/4 patch size) balances speed vs. quality
- **Gaussian weighting:** Center patch = 1.0, edges fade smoothly
- **Batch inference:** Process 16 overlapping windows at once

## Outputs
1. **Probability map:** `modeling/outputs/inference/nyc_2022/probability_map.tif`
   - GeoTIFF, float32, [0, 1], same CRS/geotransform as NAIP
   - ~1 GB file

2. **Binary threshold:** `...binary_th05.tif` (threshold=0.5)
3. **Post-processed:** `...binary_cleaned.tif` (small regions removed)

## Compute Cost
| Approach | Cost |
|----------|------|
| CPU inference (slow) | $0 |
| GPU (T4 on Kaggle) | $0 (within quota) |
| GPU (GCP T4, 1 run) | ~$0.35–1.05 |

Expected time: 2–4 hours on T4 (depends on I/O).

## Acceptance Criteria
- [ ] Inference script runs end-to-end
- [ ] Output GeoTIFF readable in QGIS/rasterio
- [ ] CRS matches NAIP (EPSG:32618 or similar)
- [ ] Probability values in [0, 1]
- [ ] No NaN or inf values
- [ ] Overlap blending looks smooth (no visible seams)
- [ ] Binary threshold creates clean mask
- [ ] Small region removal works (spurious isolated pixels removed)

## Visualization Sanity Checks
1. **Overlay on NAIP RGB:** Vacant pixels should be bare/sparse vegetation
2. **Known parcels:** Manually inspect a few MapPLUTO vacant parcels—do they light up?
3. **Known buildings:** Buildings should be dark (low probability)

## Parallelization
- Can run while 3.2/3.3/3.4 are in progress (doesn't depend on them)
- Use best model from 3.1 baseline run
- If 3.3 yields better hyperparams, re-run inference

## Dependencies
- rasterio (windowed I/O)
- torch (loaded model)
- scipy.ndimage (morphological operations)

## References
- Huang et al. (2018), patch-wise inference for large RS images
